src/invariant/OrderBookIsNotCrossed.cpp

```diff
@@ -0,0 +1,239 @@
+#ifdef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
+// Copyright 2019 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "invariant/OrderBookIsNotCrossed.h"
+#include "invariant/InvariantManager.h"
+#include "ledger/LedgerTxn.h"
+#include "lib/util/format.h"
+#include "main/Application.h"
+#include "xdr/Stellar-ledger-entries.h"
+
+namespace stellar
+{
+
+std::vector<std::pair<Asset, Asset>>
+extractAssetPairs(Operation const& op, LedgerTxnDelta const& ltxd)
+{
+    switch (op.body.type())
+    {
+    case MANAGE_BUY_OFFER:
+    {
+        auto const& offer = op.body.manageBuyOfferOp();
+        return {std::pair<Asset, Asset>(offer.selling, offer.buying)};
+    }
+    case MANAGE_SELL_OFFER:
+    {
+        auto const& offer = op.body.manageSellOfferOp();
+        return {std::pair<Asset, Asset>(offer.selling, offer.buying)};
+    }
+    case CREATE_PASSIVE_SELL_OFFER:
+    {
+        auto const& offer = op.body.createPassiveSellOfferOp();
+        return {std::pair<Asset, Asset>(offer.selling, offer.buying)};
+    }
+    case PATH_PAYMENT:
+    {
+        auto const& pp = op.body.pathPaymentOp();
+
+        // if no path, only have a pair between send and dest
+        if (pp.path.size() == 0)
+        {
+            return {std::pair<Asset, Asset>(pp.sendAsset, pp.destAsset)};
+        }
+
+        std::vector<std::pair<Asset, Asset>> assets;
+        // For send, dest, {A, B} we get: {send, A}, {A, B}, {B, dest}
+        for (int i = 0; i < pp.path.size(); ++i)
+        {
+            // beginning: send -> A
+            if (i == 0)
```

Comment:
is does not make sense for those conditions to be inside loop
you can loop from 1 to pp.path.size() - 1 and have those special cases outside of loop

Reply:
Good point. Fixed, simpler now.